crates/cxx-qt-gen/src/parser/naming.rs

```diff
@@ -0,0 +1,507 @@
+// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::collections::BTreeMap;
+
+use quote::format_ident;
+use syn::{
+    spanned::Spanned, token::Brace, Attribute, Ident, Item, ItemEnum, ItemForeignMod, ItemStruct,
+    Path, Result,
+};
+
+use crate::syntax::{
+    attribute::attribute_find_path, expr::expr_to_string,
+    foreignmod::foreign_mod_to_foreign_item_types, path::path_from_idents,
+};
+
+use super::{cxxqtdata::ParsedCxxQtData, qenum::ParsedQEnum, qobject::ParsedQObject};
+
+/// The purpose of this struct is to store all nameable types.
+///
+/// This is used by the generator phase to find types by their identifier, such that they can be
+/// fully qualified in Rust and C++.
+#[derive(Default, Debug)]
+pub struct TypeNames {
+    /// Map of the cxx_name of any types defined in CXX extern blocks
+    ///
+    /// This is used in the C++ generation to map the Rust type name to the C++ name
+    pub cxx_names: BTreeMap<String, String>,
+    /// Map of the namespace of any types or methods defined in CXX extern blocks
+    ///
+    /// This is used in the C++ generation to map the Rust type name to the C++ name
+    pub namespaces: BTreeMap<String, String>,
+    /// Mappings for CXX types when used outside the bridge
+    ///
+    /// This is used in the Rust generation to map the bridge type A to ffi::B
+    pub qualified: BTreeMap<Ident, Path>,
+}
+
+impl TypeNames {
+    /// The "Naming" phase.
+    /// Extract all nameable types from the CXX-Qt data and the CXX items.
+    ///
+    /// This allows the generator to fully-qualify all types in the generated code.
+    pub fn from_parsed_data(
```

Comment:
ideally later if we did go the `syn::Parse` route to a `CxxQtMod` this would then take just that as the input.

```rust
TypeNames::from_parsed_data(module: &CxxQtMod)
```

This would then mean not needing to pass in the bridge namespace or module ident or items as everything would be contained in `CxxQtMod`.

Then this struct could iterate over the required blocks to build up the cxx_names, namespaces, qualified names etc.

Reply:
That's the basic idea.
Parse everything first, then extract the names from this parsed info, whatever that looks like.